sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -218,6 +280,49 @@ object Hash {
             acc.addTypedValue(x._1).addTypedValue(x._2)
           )
       }
+
+    private def addTypedValue(fieldNumber: Int, value: Value): this.type =
+      value match {
+        case Value.ValueUnit =>
+          this
+        case Value.ValueBool(b) =>
+          add(fieldNumber, b)
+        case Value.ValueInt64(v) =>
+          add(fieldNumber, v)
+        case Value.ValueNumeric(v) =>
+          add(fieldNumber, v)
+        case Value.ValueTimestamp(v) =>
+          add(fieldNumber, v.micros)
+        case Value.ValueDate(v) =>
+          add(fieldNumber, v.days)
+        case Value.ValueParty(v) =>
+          // No default
+          add(fieldNumber).add(v)
+        case Value.ValueText(v) =>
+          add(fieldNumber, v)
+        case Value.ValueContractId(cid) =>
+          // No default
+          add(fieldNumber).addCid(cid)
+        case Value.ValueOptional(opt) =>
+          iterateOver(fieldNumber, opt)(_ addTypedValue _)
+        case Value.ValueList(xs) =>
+          iterateOver(fieldNumber, xs.toImmArray)(_ addTypedValue _)
+        case Value.ValueTextMap(xs) =>
+          iterateOver(fieldNumber, xs.toImmArray)((acc, x) => acc.add(x._1).addTypedValue(x._2))
+        case Value.ValueRecord(_, fs) =>
+          // No default
+          add(fieldNumber).addRecord(fs)
+        case Value.ValueVariant(_, variant, v) =>
+          // No default
+          add(fieldNumber).add(variant).addTypedValue(v)
+        case Value.ValueEnum(_, v) =>
+          // No default
+          add(fieldNumber).add(v)
+        case Value.ValueGenMap(entries) =>
+          iterateOver(fieldNumber, entries.iterator, entries.length)((acc, x) =>
+            acc.addTypedValue(x._1).addTypedValue(x._2)
```

Comment:
This too suffers from collisions because records are not properly delineated. All the key-value pairs are merely converted to bytes and concatenated. Since records with default-only fields disappear, you get hash collisions.